basalt-server-lib/src/server/teams.rs

```diff
@@ -0,0 +1,193 @@
+use bedrock::Config;
+use chrono::Utc;
+use dashmap::DashMap;
+use serde::{Deserialize, Serialize};
+
+use crate::repositories::users::Username;
+
+#[derive(Debug, PartialEq, Eq, Default, Copy, Clone, Deserialize, Serialize, utoipa::ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct TeamInfo {
+    /// When the team last contacted the server
+    pub last_seen: Option<chrono::DateTime<Utc>>,
+    /// Whether or not the team has checked into the competition by logging in
+    pub checked_in: bool,
+    /// Just a flag stating whether or not the team has deliberately disconnected
+    pub disconnected: bool,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Deserialize, Serialize, utoipa::ToSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct TeamFull {
+    /// Username of team/player
+    pub team: Username,
+    /// Contains full information about team
+    #[serde(flatten)]
+    pub info: TeamInfo,
+}
+
+impl TeamInfo {
+    fn check(&mut self) {
+        self.checked_in = true;
+        self.last_seen = Some(Utc::now());
+        self.disconnected = false;
+    }
+    fn disconnect(&mut self) {
+        self.disconnected = true;
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, serde::Serialize, utoipa::ToSchema)]
+pub struct TeamWithScore {
+    pub score: f64,
+    #[serde(flatten)]
+    pub team_info: TeamFull,
+}
+
+pub struct TeamManagement {
+    teams: DashMap<Username, TeamInfo>,
+}
+
+impl TeamManagement {
+    pub fn from_config(cfg: &Config) -> Self {
+        let teams: DashMap<Username, TeamInfo> = DashMap::new();
+        for t in &cfg.accounts.competitors {
+            teams.insert(t.name.clone().into(), TeamInfo::default());
+        }
+        TeamManagement { teams }
+    }
+
+    pub fn check_in(&self, name: &Username) {
+        self.teams.entry(name.clone()).and_modify(|t| t.check());
+    }
+
+    pub fn disconnect(&self, name: &Username) {
+        self.teams
+            .entry(name.clone())
+            .and_modify(|t| t.disconnect());
+    }
+
+    pub fn list(&self) -> Vec<TeamFull> {
+        self.teams
+            .clone()
+            .into_read_only()
+            .iter()
+            .map(|(k, v)| TeamFull {
+                team: k.clone(),
+                info: *v,
+            })
+            .collect::<Vec<TeamFull>>()
+    }
```

Comment:
Two things about this:
- If you're cloning the map, you don't need to also clone the keys.  You can use `into_iter` to own the keys and values.
- Since this function is only every used as an iterator, just return an iterator so we never allocate a vec when we don't need it

```suggestion
    pub fn list(&self) -> impl Iterator<Item = TeamFull> {
        self.teams
            .clone()
            .into_iter()
            .map(|(k, v)| TeamFull {
                team: k,
                info: v,
            })
    }
```

Reply:
Good idea. Done